docs/proposals/20240930-machine-drain-rules.md

```diff
@@ -0,0 +1,324 @@
+
+* `cluster.x-k8s.io/drain: skip`
+* `cluster.x-k8s.io/drain-order: <order>`
+
+#### Node drain behavior
+
+The following describes the new algorithm that decides which Pods should be drained and in which order.
+
+* List all Pods on the Node
+* For each Pod (first match applies):
+  * If the Pod is a DaemonSet Pod or a static Pod
+    * \=\> use `behavior: Skip`
+  * If the Pod has `cluster.x-k8s.io/drain: skip`
+    * \=\> use `behavior: Skip`
+  * if the Pod has `cluster.x-k8s.io/drain-order: <order>`
+    * \=\> use `behavior: Drain` and `order: <order>`
+  * If there is a matching `MachineDrainRule`
+    * \=\> use `behavior` and `order` from the first matching `MachineDrainRule`
+  * Otherwise:
+    * \=\> use `behavior: Drain` and `order: 0`
+* If there are no more Pods to be drained
+  * \=\> drain is completed
+* Else:
+  * \=\> evict all Pods with the lowest `order`, update condition and requeue
+
+Notes:
+
+* It does not make sense to evict DaemonSet Pods, because the DaemonSet controller always [adds tolerations for the 
+  unschedulable taint to all its Pods](https://kubernetes.io/docs/concepts/workloads/controllers/daemonset/#taints-and-tolerations)
+  so they would be just immediately rescheduled.
+* It does not make sense to evict static Pods, because the kubelet won’t actually terminate them and the Pod object 
+  will get recreated.
+* If multiple `MachineDrainRules` match the same Pod, the “first” rule (based on alphabetical order) will be applied
+* `Pods` with drain behavior `Drain` and given order will be evicted only after all other `Pods` with a lower order 
+  have been already deleted
+* When waiting for volume detach, volumes belonging to `Pods` with drain behavior `Skip` are going to be ignored.
+  Or in other words, Machine deletion can continue if only volumes belonging to `Pods` with drain behavior `Skip`
+  are still attached to the Node.
+
+#### Changes to wait for volume detach
+
+Today, after Node drain we are waiting for **all** volumes to be detached. We are going to change that behavior to ignore
+all attached volumes that belong to Pods for which we skipped the drain.
+
+### Security Model
+
+This proposal will add a new `MachineDrainRule` CRD. The Cluster API core controller needs permissions to read this CRD.
+While permissions to write the CRD would be typically given to the cluster admin (the persona who has permissions to manage
+Cluster CRs), it is also possible to further restrict access.
+
+### Risks and Mitigations
+
+* There is ongoing work in Kubernetes in this area that we should eventually align with ([KEP-4212](https://github.com/kubernetes/enhancements/issues/4212),
+  [KEP-4563](https://github.com/kubernetes/enhancements/issues/4563))
+  * We are going to provide feedback to the KEPs to try to make sure the same use cases as in the current proposal are covered.
+
+## Alternatives
+
+**Add Node drain configuration to the Machine object**
```

Comment:
have we considered adding this to the scalable resource intent, MD/MS?

Reply:
I think it is implicitly yes, because by adding to the machine spec it also gets added to MD/MS template specs (and the in CC and Cluster topology as we always do).

This would have been a lot of duplication for setting which seems orthogonal to machine management.

Reply:
Made it slightly clearer that this option also includes MD / MS